packages/application/src/lab.ts

```diff
@@ -0,0 +1,404 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { PageConfig, URLExt } from '@jupyterlab/coreutils';
+
+import { Base64ModelFactory } from '@jupyterlab/docregistry';
+
+import { IRenderMime } from '@jupyterlab/rendermime-interfaces';
+
+import { Token } from '@phosphor/coreutils';
+
+import { DisposableDelegate, IDisposable } from '@phosphor/disposable';
+
+import { ISignal, Signal } from '@phosphor/signaling';
+
+import { JupyterFrontEnd, JupyterFrontEndPlugin } from './frontend';
+
+import { createRendermimePlugins } from './mimerenderers';
+
+import { ILabShell, LabShell } from './shell';
+
+/* tslint:disable */
+/**
+ * The application status token.
+ */
+export const ILabStatus = new Token<ILabStatus>(
+  '@jupyterlab/application:ILabStatus'
+);
+/* tslint:enable */
+
+/**
+ * An interface for JupyterLab-like application status functionality.
+ */
+export interface ILabStatus {
+  /**
+   * A signal for when application changes its busy status.
+   */
+  readonly busySignal: ISignal<JupyterFrontEnd, boolean>;
+
+  /**
+   * A signal for when application changes its dirty status.
+   */
+  readonly dirtySignal: ISignal<JupyterFrontEnd, boolean>;
+
+  /**
+   * Whether the application is busy.
+   */
+  readonly isBusy: boolean;
+
+  /**
+   * Whether the application is dirty.
+   */
+  readonly isDirty: boolean;
+
+  /**
+   * Set the application state to busy.
+   *
+   * @returns A disposable used to clear the busy state for the caller.
+   */
+  setBusy(): IDisposable;
+
+  /**
+   * Set the application state to dirty.
+   *
+   * @returns A disposable used to clear the dirty state for the caller.
+   */
+  setDirty(): IDisposable;
+}
+
+/**
+ * JupyterLab is the main application class. It is instantiated once and shared.
+ */
+export class JupyterLab extends JupyterFrontEnd<ILabShell>
+  implements ILabStatus {
```

Comment:
It seems odd to me to implement `ILabStatus` directly on the application and then return `app` for the `ILabStatus` plugin above. I would think that we could keep that as a separate class and reduce the `JupyterLab` API surface area further. What is the rationale for keeping it here?

Reply:
Well, it effectively *is* what you describe because if you import the `ILabStatus` token, you have no idea that you're actually getting `JupyterLab`, at least in TypeScript. Do you think that's insufficient separation? If we change it to a separate class, it'll be totally transparent to users from the way it currently is implemented.

Reply:
I think it's sufficient from the perspective of users/extension devs. If we separate it further it could better insulate *us* from more refactoring of the top-level application.